plugin/storage/badger/spanstore/cache.go

```diff
@@ -90,67 +102,99 @@ func (c *CacheStore) loadOperations(service string) {
 		}
 		return nil
 	})
 }
 
 // Update caches the results of service and service + operation indexes and maintains their TTL
-func (c *CacheStore) Update(service, operation string, expireTime uint64) {
+func (c *CacheStore) Update(service, operation string, kind model.SpanKind, expireTime uint64) {
 	c.cacheLock.Lock()
-
 	c.services[service] = expireTime
 	if _, ok := c.operations[service]; !ok {
-		c.operations[service] = make(map[string]uint64)
+		c.operations[service] = make(map[model.SpanKind]map[string]uint64)
+	}
+	if _, ok := c.operations[service][kind]; !ok {
+		c.operations[service][kind] = make(map[string]uint64)
 	}
-	c.operations[service][operation] = expireTime
+	c.operations[service][kind][operation] = expireTime
 	c.cacheLock.Unlock()
 }
 
 // GetOperations returns all operations for a specific service & spanKind traced by Jaeger
-func (c *CacheStore) GetOperations(service string) ([]spanstore.Operation, error) {
-	operations := make([]string, 0, len(c.services))
+func (c *CacheStore) GetOperations(service string, kind string) ([]spanstore.Operation, error) {
+	operations := make([]spanstore.Operation, 0, len(c.services))
 	//nolint: gosec // G115
 	t := uint64(time.Now().Unix())
 	c.cacheLock.Lock()
 	defer c.cacheLock.Unlock()
-
 	if v, ok := c.services[service]; ok {
 		if v < t {
 			// Expired, remove
 			delete(c.services, service)
 			delete(c.operations, service)
 			return []spanstore.Operation{}, nil // empty slice rather than nil
 		}
-		for o, e := range c.operations[service] {
-			if e > t {
-				operations = append(operations, o)
-			} else {
-				delete(c.operations[service], o)
+		if kind != "" {
+			spanKind, err := model.SpanKindFromString(kind)
+			if err != nil {
+				return nil, err
+			}
+			for o, e := range c.operations[service][spanKind] {
+				operations = insertOperations(c, operations, service, o, spanKind, t, e)
+				sort.Slice(operations, func(i, j int) bool {
+					return operations[i].Name < operations[j].Name
+				})
+			}
+		} else {
+			for sKind := range c.operations[service] {
+				for o, e := range c.operations[service][sKind] {
+					operations = insertOperations(c, operations, service, o, sKind, t, e)
+					sort.Slice(operations, func(i, j int) bool {
+						if operations[i].SpanKind == operations[j].SpanKind {
+							return operations[i].Name < operations[j].Name
+						}
+						return operations[i].SpanKind < operations[j].SpanKind
+					})
+				}
 			}
 		}
 	}
+	return operations, nil
+}
 
-	sort.Strings(operations)
-
-	// TODO: https://github.com/jaegertracing/jaeger/issues/1922
-	// 	- return the operations with actual spanKind
-	result := make([]spanstore.Operation, 0, len(operations))
-	for _, op := range operations {
-		result = append(result, spanstore.Operation{
-			Name: op,
-		})
+func insertOperations(c *CacheStore, operations []spanstore.Operation, service, operation string, kind model.SpanKind, t, e uint64) []spanstore.Operation {
```

Comment:
* please add a description to this function, I don't understand what it's doing (and it has `delete` inside it while being called `insert`).
* spell `t, e` args, it's impossible to tell their mening